library/foreman_provisioner.py

```diff
@@ -0,0 +1,292 @@
+     default: true
+     required: false
+'''
+
+
+MIN_SUPPORTED_VERSION = 2
+MGMT_SUPPORTED_STRATEGIES = ['foreman', 'ipmi']
+WAIT_TO_FINISH_BUILDING = 10
+WAIT_TO_FINISH_BOOTING = 10
+
+
+class ForemanManager(object):
+    """
+    This class represents a simple interface for foreman* to easily rebuild /
+    get / reserve hosts from foreman.
+    *Foreman: http://theforeman.org/
+    """
+    def __init__(self, url, username, password, extra_headers=None, version=2):
+        """
+        :param url: the url of the foreman we wish to authenticate with
+        :param username: the username we will use to login
+        :param password: the password we will use to login
+        :param extra_headers: if we require extra headers to be added to the
+        http request
+        :param version: the version of foreman API we wish to use (default: 2)
+        :type version: int
+        """
+        if version < MIN_SUPPORTED_VERSION:
+            raise Exception("API version: {0} "
+                            "is not supported at the moment".format(version))
+
+        self.session = requests.Session()
+        self.session.auth = (username, password)
+
+        headers = {'Accept': 'application/json',
+                   'Content-type': 'application/json'}
+
+        if extra_headers:
+            headers.update(extra_headers)
+
+        self.session.headers.update(headers)
+
+        self.url = url.rstrip('/')
+        self.default_uri = '/api/v2/hosts/'
+
+    def reserve_host(self, host_id):
+        """
+        This method 'tags' a host as reserved in foreman
+        :param host_id: the name of ID of the host we wish to reserve
+        :returns: the host information on success, else empty body
+        :rtype: list of dictionaries -- [{"host": {}}]
+        """
+        #TODO(tkammer): add the option to provide the query itself after "?"
+        request_url = '{0}/api/hosts_reserve' \
+                      '?query=name ~ {1}'.format(self.url, host_id)
+        response = self.session.get(request_url, verify=False)
+        body = response.json()
+        return body
+
+    def release_host(self, host_id):
+        """
+        This method removed the 'tag' made by 'reserve_host" in foreman
+        :param host_id: the name or ID of the host we wish to release
+        :returns: the host name
+        :rtype: list of strings
+        """
+        request_url = '{0}/api/hosts_release' \
+                      '?query=name ~ {1}'.format(self.url, host_id)
+        response = self.session.get(request_url, verify=False)
+        body = response.json()
+        Exception(body)
+        return body
+
+    def get_host(self, host_id):
+        """
+        This method returns the host details as listed in the foreman
+        :param host_id: the name or ID of the host we wish to get
+        :returns: host information
+        :rtype: dict
+        """
+        request_url = '{0}/{1}/{2}'.format(self.url, self.default_uri, host_id)
```

Comment:
Please use 'urlparse.urljoin' function instead of format string in all methods where you construct the 'request_url'